### PR TITLE
upgrades: delete tenants override for the `version` key

### LINKED ISSUE #125702: multitenant: bogus row in `system.tenant_settings` may prevent tenant upgrade
**TLDR**:

Clusters that upgraded to 23.1 using a version older than v23.1.9 will forever have a row in the `system.tenant_settings` table for `tenant_id=0` and `version=23.1`. This may prevent tenants from upgrading in the future as they might have an incorrect view of the storage cluster's version in specific circumstances.

*****

### Long winded background

In the 23.1 release cycle, we introduced a startup migration[^1] (later rewritten as a permanent upgrade) that added a row to the `system.tenant_settings` table with the storage cluster's version information. Tenants would get notified of changes in this table (via the `settingswatcher`) in order to keep an updated view of the storage cluster's version.

However, the startup migration had a bug in its implementation; specifically, it used the `%d` 'verb' to generate the query, but the `tenantID` variable is of type `*int64`[^2]:

https://github.com/cockroachdb/cockroach/blob/fd42d376fb5ec42e9d9a30fcdcb210a8a79ef0e2/pkg/startupmigrations/migrations.go#L760-L770

Similar logic was added when changing the `version` setting in the system tenant; the version would be updated accordingly in this new row in `system.tenant_settings`:

https://github.com/cockroachdb/cockroach/blob/fd42d376fb5ec42e9d9a30fcdcb210a8a79ef0e2/pkg/sql/set_cluster_setting.go#L558-L569

Note, however, that _this_ implementation doesn't have same bug: while we still use `tree.NewDInt(0)`, we're using the `$1` placeholder instead of `fmt.Sprintf`.

The bug with the permanent upgrade was eventually noticed (#105858) and fixed (#105854). Shortly after, the way tenants learn about changes in the storage cluster's version also changed to no longer rely on an entry in the `system.tenant_settings` table: #106037. Both of these fixes were backported to the 23.1 release branch and were available in public releases starting from the v23.1.9 version.

A corollary to all of this is that the state of the `system.tenant_settings` table is **different depending on whether the cluster was bootstrapped on 23.1 vs bootstrapped on an older version and upgraded to 23.1**.

<details>
<summary>Bootstrapping on 23.1</summary>

```console
$ roachprod create local -n1
$ roachprod stage local release v23.1.8
$ roachprod start local
$ roachprod sql local:1 -- -e "SELECT * FROM system.tenant_settings;"
    tenant_id   |  name   |               value               |        last_updated        | value_type | reason
----------------+---------+-----------------------------------+----------------------------+------------+---------
  1374467842288 | version | \x12\b\b\x17\x10\x01\x18\x00 \x00 | 2024-06-14 18:37:22.956285 | m          | NULL
(1 row)
```
</details>

<details>
<summary>Upgrading to 23.1</summary>

```console
$ roachprod create local -n1
$ roachprod stage local release v22.2.19
$ roachprod start local
$ roachprod deploy local release v23.1.8
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING version = '23.1';" # wait for migrations
$ roachprod sql local:1 -- -e "SELECT * FROM system.tenant_settings;"
  tenant_id |  name   |               value               |        last_updated        | value_type | reason
------------+---------+-----------------------------------+----------------------------+------------+---------
          0 | version | \x12\b\b\x17\x10\x01\x18\x00 \x00 | 2024-06-14 18:43:23.225887 | m          | NULL
(1 row)
```
</details>

As can be seen, the `tenant_id` column is `0` only when upgrading go 23.1; this is relevant because `0` encodes "all-tenants" overrides[^3].

### The bug

The fix introduced in #106037 (and present in v23.1.9+) works around potentially wrong entries in the `system.tenant_settings` table by "injecting" the real storage cluster version when starting up:

https://github.com/cockroachdb/cockroach/blob/a62d88b12af22f8eeecdae1f0f095777241e47e8/pkg/server/node.go#L2071-L2077

This works reasonably well; however, there are two other situations where the tenant's view of cluster settings (including the storage cluster's version) change (or rather, is overwritten):

* when a `SystemVisible` cluster setting is updated[^4]
* when a cluster setting is updated for _all tenants_[^5]

Since, in these cases, the `version` entry is not being overwritten, the tenant will update its local cache with whatever is present in the `system.tenant_settings` table (which will be outdated if the cluster has upgraded from 23.1 since).

It's easy to reproduce this situation with roachprod (note that the script sets up a shared-process tenant but the same issue exists for separate-process tenants):

<details>
<summary>Reproduction</summary>

```console
$ roachprod create local -n1
$ roachprod stage local release v22.2.19
$ roachprod start local
$ roachprod deploy local release v23.1.8
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING version = '23.1';" # wait for migrations
$ roachprod start-sql my-tenant --storage-cluster local # create shared-process tenant
$ roachprod deploy local release v23.2.5 # upgrade to 23.2
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING version = '23.2';" # wait for migrations
$ roachprod sql local:1 -- -e "SET CLUSTER SETTING kv.closed_timestamp.target_duration = '5s';" # set an arbitrary SystemVisible setting
$ roachprod sql local:1 --cluster my-tenant -- -e "SET CLUSTER SETTING version = '23.2';" # attempt to upgrade tenant
```

The last command in that script fails with:

```
ERROR: preventing tenant upgrade from running as the storage cluster has not yet been upgraded: storage cluster version = 23.1, tenant cluster version = 23.1
```

which is wrong since we just upgraded the cluster to 23.2 two commands above.

Note: this uses the "updating a `SystemVisible` setting" approach to reproducing the bug. We could also change the default for all tenants with something like `"ALTER VIRTUAL CLUSTER ALL SET CLUSTER SETTING {ApplicationLevel setting};"` and the result would be the same.
</details>

### Possible solutions

I can imagine at least a few ways we could go about dealing with this situation:

* Write a migration to delete the `system.tenant_settings` entry where `tenant_id=0` and `name=version`. To me this would be ideal as it would definitely clear up the bogus row in that table. The fact that this wasn't done in the 23.2 cycle may suggest this could have undesired consequences.

* Apply the same "injection" of the storage cluster version done at startup in the two situations mentioned above (changing `SystemVisible` and "all-tenants" defaults). Not as clean, and the bug could be reintroduced if we add some other way to refresh the tenant's view of these settings.

* Reject regressions in the storage cluster version in the code below, as that would indicate a programming error. This is also not ideal as it could mask other issues:

https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/settingswatcher/settings_watcher.go#L514-L515

* Do nothing and if anyone finds themselves in this situation, ask them to restart the nodes, as that will cause the startup logic to kick in and the right storage cluster version will be cached. Unsatisfying IMO.

[^1]: see #91071 for details.
[^2]: https://github.com/cockroachdb/cockroach/blob/fd42d376fb5ec42e9d9a30fcdcb210a8a79ef0e2/pkg/sql/sem/tree/datum.go#L689-L691
[^3]: for definition and uses, see: https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/tenantsettingswatcher/row_decoder.go#L34
[^4]: https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/tenantsettingswatcher/watcher.go#L270-L282
[^5]: https://github.com/cockroachdb/cockroach/blob/1ed54694b792af3d9c90961fb32e690adb2a1325/pkg/server/node.go#L2459-L2466

Jira issue: CRDB-39563

### PR DESCRIPTION
**Pull request by renatolabs**
**sql: disallow setting version setting key for any tenant**
This commit disallows callers from setting the `version` key as an
override for any tenants as below:

```
ALTER VIRTUAL CLUSTER ALL SET CLUSTER SETTING version = 'NN.N';
```

or:

```
ALTER VIRTUAL CLUSTER "some-tenant" SET CLUSTER SETTING version = 'NN.N';
```

The storage cluster's version is managed automatically and allowing
overrides for tenants is a footgun.

**upgrades: delete tenants override for the `version` key**
Clusters that upgraded to `23.1` using a release older than `v23.1.9`
will have an override for the `version` key that applies to all
tenants: in other words, a row in the `system.tenant_settings` table
where the `tenant_id` column is `0`.

Starting from `v23.1.9`, however, tenants learn about changes in the
storage cluster's version using a different mechanism and no longer
rely on this all-tenants override. However, the existence of this row
might prevent tenant upgrades in certain scenarios when the tenant's
cluster settings cache is overwritten.

In this commit, we add a migration to delete *any* tenant overrides
for the `version` key (whether they apply to all tenants or not). The
`version` key shouldn't be set for tenants that way, so having that
row lingering in the settings table can lead to bugs in the future.

**server: inject cluster version when changing system visible settings**
This commit adds the storage cluster's version when notifying tenants
of changes in `SystemVisible` cluster settings. This change is to work
around the situation where the `system.tenant_settings` table has
wrong data, in which case the tenant could have the wrong view of the
storage cluster's version, potentially preventing upgrades.

Fixes: #125702

Release note (bug fix): Fix a bug in non-system tenants where tenants
could have an incorrect view of the storage cluster's version in certain
situations, preventing tenant upgrades until they were restarted.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/cluster_settings
Comment (stevendanna) [inline pkg/sql/logictest/testdata/logic_test/cluster_settings:306]: I know I raised this use case, but I wonder if it is actually ever going to happen.  Perhaps it is safe to disallow it until we have a real use case.

File: pkg/upgrade/upgrades/upgrades.go
Comment (stevendanna) [inline pkg/upgrade/upgrades/upgrades.go:140]: We could remove this erroneous value on restore in `tenantSettingsTableRestoreFunc`.

**Review by stevendanna — COMMENTED** [review]
Reviewed the draft since I'll be out next week and wanted to see your approach.

Thanks for taking this on!

File: pkg/server/node.go
Comment (renatolabs) [reply pkg/server/node.go:2442]: The main reason was that this is the mechanism used during startup and when a `SystemVisible` cluster setting changes, so it made sense to me to do the same thing here.

File: pkg/sql/logictest/testdata/logic_test/cluster_settings
Comment (renatolabs) [reply pkg/sql/logictest/testdata/logic_test/cluster_settings]: Yeah, I had the same thought, especially there's still a workound if for some reason this was necessary (using a direct `INSERT` syntax instead of `ALTER VIRTUAL CLUSTER`). Updated.

File: pkg/upgrade/upgrades/upgrades.go
Comment (renatolabs) [reply pkg/upgrade/upgrades/upgrades.go]: Pushed something. Where are changes like this tested?

Comment (stevendanna) [reply pkg/upgrade/upgrades/upgrades.go]: The easiest way is probably to add a new  backup artifact in

```
pkg/ccl/backupccl/testdata/restore_old_versions/
```

You'd create a cluster with the "bad" state and then create a backup for it to `nodelocal://` then you can copy the backup files into a directory there and then modify 
`pkg/ccl/backupccl/restore_old_versions_test.go` to include your new test case.

**Comment by renatolabs** [comment]
Other than the unit tests in the PR, I also verified manually that an upgrade to this branch deletes the bad data; also verified that cherry-picking the 3rd commit in this PR on 23.2 and 24.1 fixes the issue reported using the reproduction steps in #125702.

File: pkg/upgrade/upgrades/upgrades.go
Comment (renatolabs) [reply pkg/upgrade/upgrades/upgrades.go]: Thanks, done.

**Comment by renatolabs** [comment]
Another friendly reminder -- could I get a review on this PR please? Maybe @dt, since Steven is out this week?

**Review by RaduBerinde — APPROVED** [review]
I don't grasp all the background details of how we got here, but the code changes LGTM.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125881#-:-O0LwxwL66dgwpetJguI:b-9ls1xq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, @renatolabs, and @stevendanna)

___
*[`pkg/server/settingswatcher/settings_watcher.go` line 654 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125881#-O0Lwdbz2lbYCJDICrja:-O0Lwdbz2lbYCJDICrjb:b3ni5z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0458b9e0af862c42ae8eedad46f69992c2ffe32/pkg/server/settingswatcher/settings_watcher.go#L654)):*
> ```Go
> 	// TODO(multitenant): remove this logic once the minimum supported
> 	// version is 24.1+.
> 	overrides := append(
> ```

[nit] can just be
```
[]kvpb.TenantSetting{
  payload,
  {
    InternalKey: ..,
    ..
  },
}
```

**Review by stevendanna — APPROVED** [review]
Overall this make sense to me. Marking approved, but I do think it would be nice to clean up some of the unnecessary test files before merge so they don't live forever in the git history.

I think the plan is to just backport the commit that makes the version ingestion consistent. 

Re @dt's comment on filtering the version key rather than injecting it. I do think that seems like it would be a little cleaner, but after a brief read through the settings watcher it wasn't clear to me the best place to do it (certainly the _easiest_ place to verify we are doing it correctly would be to to iterate the array and remove any version key we find).  I personally think I'm OK with what you have here since this is what you've spent time testing.

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125881)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @msbutler, @RaduBerinde, and @stevendanna)

___
*[`pkg/server/settingswatcher/settings_watcher.go` line 654 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/125881#-O0Lwdbz2lbYCJDICrja:-O0PAvSU8huRk8bT1dcM:bc8flwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0458b9e0af862c42ae8eedad46f69992c2ffe32/pkg/server/settingswatcher/settings_watcher.go#L654)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] can just be
```
[]kvpb.TenantSetting{
  payload,
  {
    InternalKey: ..,
    ..
  },
}
```
</blockquote></details>

Thanks, done

File: pkg/upgrade/upgrades/v24_2_delete_all_tenants_version_tenant_settings.go
Comment (renatolabs) [reply pkg/upgrade/upgrades/v24_2_delete_all_tenants_version_tenant_settings.go]: Yeah, makes sense to me. Migration updated.

**Comment by renatolabs** [comment]
> I think the plan is to just backport the commit that makes the version ingestion consistent.

By "version ingestion", do you mean the `restore` logic part of this PR? If so, wouldn't 23.2 and 24.1 still be susceptible to the issue we're trying to fix?

> but after a brief read through the settings watcher it wasn't clear to me the best place to do it

Same. I spent some time trying to find the right place to put this logic in the settings watcher, and didn't feel very confident that the change wouldn't break something inadvertently. Since all this injection logic has a removal timer on it, I figured it wouldn't be too bad to do it in `node.go`.

That said, if you feel like we should do something else (e.g., filter out by scanning the overrides as you suggested), I'm happy to make that change.

**Comment by stevendanna** [comment]
> By "version ingestion", do you mean the restore logic part of this PR? If so, wouldn't 23.2 and 24.1 still be susceptible to the issue we're trying to fix?

Sorry, I meant to type "version injection" here.

**Comment by renatolabs** [comment]
Did another batch of testing of these changes after recent reviews. Everything still looks good, so I'm merging.

TFTRs!

bors r=RaduBerinde,stevendanna

**Pull request merged**
